hooks/bootstrap.py

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) 2018 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import contextlib
+
+from sgtk import get_hook_baseclass
+
+
+class BootstrapHook(get_hook_baseclass()):
+    """
+    The bootstrap hook can be added to a configuration's core hooks folder
+    (``config/core/hooks``) in order to override some of the functionality of
+    the ToolkitManager's bootstrapping process.
+
+    The hook will be instantiated only after a configuration has been selected
+    by the ``ToolkitManager``. Therefore, the ``download_bundle`` method
+    will not be invoked to download a configuration. However, the Toolkit Core,
+    applications, frameworks and engines will be downloaded through the hook.
+    """
+
+    def init(self, shotgun, pipeline_configuration_id, configuration_descriptor):
```

Comment:
perhaps add kwargs here so that we make client code forwards compatible?

Reply:
I think our clients should put kwargs in their methods, but this copy of the hook is invoked by us, we should know how many params the core is going to pass in, no? Or you're worried people may be copy pasting this to start with and then forget to add the `kwargs` as good practice?

Reply:
yes, exactly (the latter!) - was what i was thinking. We do it in quite a few places - e.g. https://github.com/shotgunsoftware/tk-core/blob/master/hooks/engine_init.py.